### PR TITLE
fix: Handle ProvisionedThroughputExceededException throttling

### PR DESCRIPTION
**Pull request by spentakota**

### CODE REVIEW

File: clientlibrary/config/kcl-config.go [hunk 1]
```diff
@@ -211,6 +212,13 @@ func (c *KinesisClientLibConfiguration) WithLogger(logger logger.Logger) *Kinesi
 	return c
 }
 
+// WithMaxRetryCount sets the max retry count in case of error.
+func (c *KinesisClientLibConfiguration) WithMaxRetryCount(maxRetryCount int) *KinesisClientLibConfiguration {
```
Comment (Devika-Shanbhag) [inline clientlibrary/config/kcl-config.go:216]:
```suggestion
func (c *KinesisClientLibConfiguration) SetMaxRetryCount(maxRetryCount int) *KinesisClientLibConfiguration {
```

Comment (spentakota) [reply clientlibrary/config/kcl-config.go:216]: I was thinking to keep the 'with' prefix in order to maintain consistency as that is how all other methods for setting KCL config properties starts: https://gitlab.eng.vmware.com/entity-data-service/inventory-consumer/-/blob/master/cmd/main.go#L240

File: clientlibrary/worker/polling-shard-consumer.go [hunk 2]
```diff
@@ -166,6 +174,14 @@ func (sc *PollingShardConsumer) getRecords() error {
 			if errors.As(err, &kmsThrottlingErr) {
 				log.Errorf("Error getting records from shard %v: %+v", sc.shard.ID, err)
 				retriedErrors++
+				// Greater than MaxRetryCount so we get the last retry
+				if retriedErrors > sc.kclConfig.MaxRetryCount {
+					log.Errorf("message", "reached max retry count getting records from shard",
```
Comment (vmwjc) [inline clientlibrary/worker/polling-shard-consumer.go:196]: Can we have a slightly different message here that maybe mentions that it's a KMS throttling error? That way we can differentiate it from the Kinesis error above.

File: clientlibrary/worker/polling-shard-consumer.go [hunk 3]
```diff
@@ -228,19 +247,57 @@ func (sc *PollingShardConsumer) waitASecond(timePassed time.Time) {
 	}
 }
 
-func (sc *PollingShardConsumer) callGetRecordsAPI(gri *kinesis.GetRecordsInput) (*kinesis.GetRecordsOutput, error) {
+func (sc *PollingShardConsumer) checkCoolOffPeriod() (error, float64) {
```
Comment (vmwjc) [inline clientlibrary/worker/polling-shard-consumer.go:250]: Conventionally, the error value is the last in the return values.

See https://gobyexample.com/errors

File: clientlibrary/worker/polling-shard-consumer.go [hunk 4]
```diff
@@ -228,19 +247,57 @@ func (sc *PollingShardConsumer) waitASecond(timePassed time.Time) {
...
+		// Wait until cool down period has passed to prevent ProvisionedThroughputExceededException
+		coolDown := float64(sc.bytesRead / MaxBytesPerSecond)
+		return maxBytesExceededError, coolDown
+	} else {
+		sc.remBytes -= sc.bytesRead
+	}
+	return nil, 0
+}
+
+func (sc *PollingShardConsumer) callGetRecordsAPI(gri *kinesis.GetRecordsInput) (*kinesis.GetRecordsOutput, error, float64) {
```
Comment (vmwjc) [inline clientlibrary/worker/polling-shard-consumer.go:274]: Conventionally, the error value is the last in the return values.

See https://gobyexample.com/errors

**Review by Devika-Shanbhag — APPROVED** [review]

**Pull request merged**